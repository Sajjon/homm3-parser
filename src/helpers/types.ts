export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export type FlaggedProp<FlagName extends string, ValueName extends string, T> =
  | Record<FlagName, true> & Record<ValueName, T>
  | Record<FlagName, false>
