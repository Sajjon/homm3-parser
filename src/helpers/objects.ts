import { ObjectType, objectTypeToMeta } from '../parsers/h3m/enums/object'
import { MetaType } from '../parsers/h3m/contracts/enums/MetaType'
import { createAdapter, TagProducer, TagOrWrapper } from 'binary-markup'

export function getMetaType(type: ObjectType): MetaType {
  if (type in objectTypeToMeta) {
    return objectTypeToMeta[type]
  }
  return MetaType.None
}

export const bitMasksArray = <T>(
  tag: TagOrWrapper<number[]>,
  bitGroups: [number, T][][],
): TagProducer<T[]> =>
  createAdapter<number[], T[]>(
    tag,
    (data: number[]) => {
      return data.reduce((acc: T[], mask, index) => {
        const bits = bitGroups[index]
        for (const [bit, skill] of bits) {
          if (mask & bit) {
            acc.push(skill)
          }
        }
        return acc
      }, [])
    },
    (skills: T[]) => {
      return bitGroups.map(bits => {
        let mask = 0
        for (const [bit, skill] of bits) {
          if (skills.includes(skill)) {
            mask |= bit
          }
        }
        return mask
      })
    },
  )
