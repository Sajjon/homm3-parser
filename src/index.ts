import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import * as pako from 'pako'
import { h3mFile } from './parsers/h3m/h3m.bml'
import { H3MFile } from './parsers/h3m/contracts'

export * from './parsers/assets'
export * from './parsers/h3m/contracts'
export * from './parsers/h3m/constants'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = pako.inflate(data)
  return parseH3M(content.buffer)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
