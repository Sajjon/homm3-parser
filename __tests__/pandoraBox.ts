import { testMap } from './helpers/utils'
import { H3MFile } from '../src'

describe('pandora box', () => {
  testMap('pandora box', 'pandora-box.h3m', (map: H3MFile) => map.objectDetails)
})
