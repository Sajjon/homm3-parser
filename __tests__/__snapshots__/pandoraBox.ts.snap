// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pandora box should parse pandora box 1`] = `
Object {
  "count": 15,
  "entries": Array [
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 1000,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "Experience",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 1,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 5,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "spell points",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 3,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": -5,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "spell points take",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 5,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 2,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "morale",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 7,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": -3,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "-morale",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 9,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 2,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "luck",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 11,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": -3,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "-luck",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 13,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              1,
              2,
              3,
              4,
              5,
              6,
              7,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": null,
          "hasGuardians": false,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 15,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              -7,
              -6,
              -5,
              -4,
              -3,
              -2,
              -1,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "take resources",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 17,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 1,
              "defense": 2,
              "knowledge": 4,
              "spellPower": 3,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "primary skills",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 19,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [
              Object {
                "level": "Basic",
                "type": "Pathfinding",
              },
              Object {
                "level": "Advanced",
                "type": "Logistics",
              },
              Object {
                "level": "Expert",
                "type": "Diplomacy",
              },
            ],
            "secondarySkillsCount": 3,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "secondary sckills",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 21,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [
              "AdmiralsHat",
            ],
            "artifactsCount": 1,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "artifacts",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 23,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [
              "AirElemental",
            ],
            "spellsCount": 1,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "spells",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 25,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [
              Object {
                "quantity": 1,
                "type": "Pikeman",
              },
            ],
            "creaturesCount": 1,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "creatures
",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 27,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVA0128.def",
        "landscapeGroup": 1,
        "metaType": "PandorasBox",
        "objectGroup": "Artifacts",
        "objectSubType": 0,
        "objectType": "PandorasBox",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "contents": Object {
            "artifacts": Array [],
            "artifactsCount": 0,
            "creatures": Array [],
            "creaturesCount": 0,
            "experience": 0,
            "luck": 0,
            "morale": 0,
            "primarySkills": Object {
              "attack": 0,
              "defense": 0,
              "knowledge": 0,
              "spellPower": 0,
            },
            "resources": Array [
              0,
              0,
              0,
              0,
              0,
              0,
              0,
            ],
            "secondarySkills": Array [],
            "secondarySkillsCount": 0,
            "spellPoints": 0,
            "spells": Array [],
            "spellsCount": 0,
          },
          "guardians": Object {
            "creatures": Array [
              Object {
                "quantity": 2,
                "type": "Archer",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 3,
                "type": "Griffin",
              },
            ],
            "hasCreatures": true,
            "message": "guardians",
          },
          "hasGuardians": true,
        },
        "type": "PandorasBox",
      },
      "oaIndex": 2,
      "x": 1,
      "y": 3,
      "z": 0,
    },
  ],
}
`;
